### PR TITLE
Add S.props

### LINKED ISSUE #295: Add an Equivalent Function to R.path
As discussed on Gitter, we should have a companion function to `S.prop`, which fulfils a similar function to [`R.path`](http://ramdajs.com/docs/#path). For internal consistency this function should be named `props`. This would be a good first contribution but if no one has picked it up over the next few days I will submit the PR.

### PR DESCRIPTION
**Pull request by svozza**
Closes #295. I wasn't sure about the text for the error message so I figured we could just bikeshed it in the PR here.

### CODE REVIEW

**Comment by svozza** [comment]
Hmmm, seems to be a SauceLabs issue:

```
23 12 2016 10:16:52.794:ERROR [launcher.sauce]: Can not start chrome 51
  Failed to start Sauce Connect:
  23 Dec 10:16:52 - Sauce Connect could not establish a connection.
23 Dec 10:16:52 - Please check your firewall and proxy settings.
23 Dec 10:16:52 - You can also use sc --doctor to launch Sauce Connect in diagnostic mode.
23 12 2016 10:16:52.796:ERROR [launcher.sauce]: Can not start firefox 47
  Failed to start Sauce Connect:
  23 Dec 10:16:52 - Sauce Connect could not establish a connection.
23 Dec 10:16:52 - Please check your firewall and proxy settings.
23 Dec 10:16:52 - You can also use sc --doctor to launch Sauce Connect in diagnostic mode.
23 12 2016 10:16:52.797:ERROR [launcher.sauce]: Can not start safari 8 (OS X 10.10)
  Failed to start Sauce Connect:
  23 Dec 10:16:52 - Sauce Connect could not establish a connection.
23 Dec 10:16:52 - Please check your firewall and proxy settings.
23 Dec 10:16:52 - You can also use sc --doctor to launch Sauce Connect in diagnostic mode.
```

**Comment by svozza** [comment]
Now the error's changed:

```
23 12 2016 10:53:24.308:ERROR [launcher.sauce]: Can not start chrome 51
  Failed to start Sauce Connect:
  Invalid Sauce Connect Credentials. 23 Dec 10:53:24 - Response: {"error": "Not authorized"}.
23 Dec 10:53:24 - Error checking overlapping tunnels.
23 Dec 10:53:24 - Cleaning up.
23 Dec 10:53:24 - Sauce Connect could not establish a connection.
23 Dec 10:53:24 - Please check your firewall and proxy settings.
23 12 2016 10:53:24.309:ERROR [launcher.sauce]: Can not start firefox 47
  Failed to start Sauce Connect:
  Invalid Sauce Connect Credentials. 23 Dec 10:53:24 - Response: {"error": "Not authorized"}.
23 Dec 10:53:24 - Error checking overlapping tunnels.
23 Dec 10:53:24 - Cleaning up.
23 Dec 10:53:24 - Sauce Connect could not establish a connection.
23 Dec 10:53:24 - Please check your firewall and proxy settings.
23 12 2016 10:53:24.310:ERROR [launcher.sauce]: Can not start safari 8 (OS X 10.10)
  Failed to start Sauce Connect:
  Invalid Sauce Connect Credentials. 23 Dec 10:53:24 - Response: {"error": "Not authorized"}.
23 Dec 10:53:24 - Error checking overlapping tunnels.
23 Dec 10:53:24 - Cleaning up.
23 Dec 10:53:24 - Sauce Connect could not establish a connection.
23 Dec 10:53:24 - Please check your firewall and proxy settings.
23 12 2016 10:53:24.310:ERROR [launcher.sauce]: Can not start safari 10 (OS X 10.11)
```

**Comment by Shard** [comment]
Pull requests should be working now, it was a configuration issue.

In regards to error messages, the one improvement I can see is moving the failed object to it's own line to give it as much space as possible, and to show it's type:

```
Invalid Path

1) {a:{b:{d:{}}}} :: Object

The value at position 1 does not have a property at 'a.b.c'
```

File: index.js [hunk 1]
```diff
@@ -2787,6 +2787,28 @@
   }
   S.prop = def('prop', {a: [Accessible]}, [$.String, a, b], prop);
 
+  //# props :: Accessible a => Array String -> a -> b
+  //.
+  //. Takes an array of property names and an object with known properties and
+  //. returns the value of the specified property. If for some reason the
+  //. object lacks the specified property, a type error is thrown.
```
Comment (davidchambers) [inline index.js]: This should be updated to reflect the fact that successive lookups are performed.

File: index.js [hunk 2]
```diff
@@ -2787,6 +2787,28 @@
...
+  //.
+  //. ```javascript
+  //. > S.props(['a', 'b', 'c'], {a: {b: {c: 1}}})
+  //. 1
+  //. ```
+  function props(path, obj) {
+    return path.reduce(function(memo, key) {
+      if (key in memo) return memo[key];
+      throw new TypeError('‘props’ expected object to have a property at ‘' +
+        path.join('.') + '’; ' + R.toString(obj) + ' does not');
```
Comment (davidchambers) [inline index.js]: I don't think we should use `path.join('.')`. It means `['foo.bar', 'baz', 'quux']` is represented as `foo.bar.baz.quux`, which is misleading. Furthermore, dot notation is not valid for all property names. I think `["a", "b", "c"]` is the most appropriate format.

**Comment by davidchambers** [comment]
Let's `s/s/S/` in the commit message and pull request title.

File: index.js (hunk 2 above)
Comment (svozza) [inline index.js]: I'm struggling to come up with the wording for this. Any suggestions?

File: index.js (hunk 1 above)
Comment (svozza) [inline index.js]: I'm struggling to come up with the wording for this. Any suggestions?

File: index.js (hunk 2 above)
Comment (svozza) [inline index.js]: :zap:

**Comment by svozza** [comment]
> In regards to error messages, the one improvement I can see is moving the failed object to it's own line to give it as much space as possible...

There error message comes from `sanctuary-def`, I only have control over string at the end.

File: index.js (hunk 1 above)
Comment (davidchambers) [inline index.js]: I suggest:

> If for some reason the object <del>lacks the specified property</del> <ins>does not have the expected structure</ins>, a type error is thrown.

Also:

> For accessing <ins>nested</ins> properties of uncertain objects, use [`gets`][1] instead.


[1]: https://sanctuary.js.org/#gets

File: index.js [hunk 3]
```diff
@@ -2787,6 +2787,28 @@
...
+  //.
+  //. ```javascript
+  //. > S.props(['a', 'b', 'c'], {a: {b: {c: 1}}})
+  //. 1
+  //. ```
+  function props(path, obj) {
+    return path.reduce(function(memo, key) {
+      if (key in memo) return memo[key];
+      throw new TypeError('‘props’ expected object to have a property at ‘' +
+        R.toString(path) + '’; ' + R.toString(obj) + ' does not');
```
Comment (davidchambers) [inline index.js]: I don't see a reason to quote the path; `[` and `]` will show clearly where it begins and ends.

Also, let's tweak the error message (and alignment) like so:

```javascript
      throw new TypeError('‘props’ expected object to have a property path ' +
                          R.toString(path) + '; ' +
                          R.toString(obj) + ' does not');
```

**Comment by davidchambers** [comment]
I rebased and resolved a merge conflict. :)

File: index.js (hunk 1 above)
Comment (davidchambers) [inline index.js]: :zap:

**Comment by davidchambers** [comment]
Could someone review this pull request? If you're near a computer, @svozza, please review my tweaks. :)

**Comment by svozza** [comment]
LGTM!

**Comment by svozza** [comment]
Another IE browser test error:

```
Can't execute code from a freed script
```

**Comment by svozza** [comment]
Seems to have fixed itself.

**Comment by davidchambers** [comment]
```
$ npm install sanctuary@0.12.0

$ node
> const S = require('sanctuary')
undefined
> S.props
props :: Accessible a => Array String -> a -> b
```

**Pull request merged**